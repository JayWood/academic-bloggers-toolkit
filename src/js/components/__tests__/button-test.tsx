import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import * as React from 'react';
import * as diff from 'snapshot-diff';

import Button from '../button';

const setup = (props: any = {}) => {
    if (!props.label) props.label = 'Click Me!';
    const component = mount(<Button {...props} />);
    return {
        component,
    };
};

describe('<Button />', () => {
    describe('Text Button', () => {
        const { component: defaultTextButton } = setup();
        const defaultJSON = toJSON(defaultTextButton);
        test('with defaults', async () => {
            expect(defaultJSON).toMatchSnapshot();
        });
        test('with focusable', () => {
            const { component } = setup({ focusable: true });
            expect(diff(defaultJSON, toJSON(component))).toMatchSnapshot();
        });
        test('with primary', () => {
            const { component } = setup({ primary: true });
            expect(diff(defaultJSON, toJSON(component))).toMatchSnapshot();
        });
        test('with flat', () => {
            const { component } = setup({ flat: true });
            expect(diff(defaultJSON, toJSON(component))).toMatchSnapshot();
        });
        test('with all variants', () => {
            const { component } = setup({ flat: true, focusable: true, primary: true });
            expect(diff(defaultJSON, toJSON(component))).toMatchSnapshot();
        });
        test('with tooltip', () => {
            const tooltip = {
                text: 'Hello world',
                position: 'top' as 'top' | 'right' | 'bottom' | 'left',
            };
            const { component } = setup({ tooltip });
            expect(diff(defaultJSON, toJSON(component))).toMatchSnapshot();
        });
    });
    describe('Icon Button', () => {
        const { component: defaultIconButton } = setup({ icon: 'plus' });
        const defaultJSON = toJSON(defaultIconButton);
        test('with defaults', async () => {
            expect(defaultJSON).toMatchSnapshot();
        });
        test('with all variants and different icon', () => {
            const { component } = setup({
                icon: 'minus',
                flat: true,
                focusable: true,
                primary: true,
            });
            expect(diff(defaultJSON, toJSON(component))).toMatchSnapshot();
        });
    });
    test('tooltip handler', () => {
        const tooltip = {
            text: 'Hello World',
            position: 'top' as 'top' | 'right' | 'bottom' | 'left',
        };
        const { component } = setup({ tooltip });
        const button = component.find('button');
        const beforeState = {
            isShowingTooltip: false,
            transform: '',
        };
        expect(component.state()).toEqual(beforeState);

        button.simulate('mouseenter');
        expect(component.state()).not.toEqual(beforeState);
        expect(component.state('isShowingTooltip')).toBeTruthy();
        expect(component.state('transform')).not.toBe('');

        button.simulate('mouseleave');
        expect(component.state()).not.toEqual(beforeState); // transform stays set
        expect(component.state('isShowingTooltip')).toBeFalsy();
        expect(component.state('transform')).not.toBe('');
    });
    test('link button handler', () => {
        window.open = jest.fn();
        const { component } = setup({ href: 'https://google.com' });
        const button = component.find('button');
        button.simulate('click');
        expect(window.open).toHaveBeenCalledWith('https://google.com', '_blank');
    });
});
