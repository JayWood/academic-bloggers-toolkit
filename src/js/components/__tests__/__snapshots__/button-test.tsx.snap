// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> Icon Button with all variants and different icon 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -1,22 +1,25 @@
 <Button
-  icon=\\"plus\\"
+  flat={true}
+  focusable={true}
+  icon=\\"minus\\"
   label=\\"Click Me!\\"
+  primary={true}
   type=\\"button\\"
 >
   <div>
     <button
       aria-describedby={undefined}
       aria-label=\\"Click Me!\\"
-      className=\\"btn-icon\\"
+      className=\\"focusable btn-primary btn-flat btn-icon\\"
       onClick={undefined}
       onMouseEnter={undefined}
       onMouseLeave={undefined}
       type=\\"button\\"
     >
       <span
-        className=\\"dashicons dashicons-plus\\"
+        className=\\"dashicons dashicons-minus\\"
       />
     </button>
     <style
       jsx={true}
     >"
`;

exports[`<Button /> Icon Button with defaults 1`] = `
<Button
  icon="plus"
  label="Click Me!"
  type="button"
>
  <div>
    <button
      aria-describedby={undefined}
      aria-label="Click Me!"
      className="btn-icon"
      onClick={undefined}
      onMouseEnter={undefined}
      onMouseLeave={undefined}
      type="button"
    >
      <span
        className="dashicons dashicons-plus"
      />
    </button>
    <style
      jsx={true}
    >
      
                    button {
                        -webkit-appearance: none;
                        -webkit-touch-callout: none;
                        -webkit-user-select: none;
                        -khtml-user-select: none;
                        -moz-user-select: none;
                        -ms-user-select: none;
                        user-select: none;
                        background: white;
                        box-sizing: border-box;
                        color: #555555;
                        background-image: none;
                        border: none;
                        border-radius: 2px;
                        box-shadow: 0 1px 3px rgba(0, 0, 0, .12), 0 1px 2px rgba(0, 0, 0, .24);
                        cursor: pointer;
                        display: inline-block;
                        font-size: 13px;
                        font-weight: 500;
                        height: 36px;
                        line-height: 36px;
                        min-width: 64px;
                        outline: 0;
                        padding: 0 16px;
                        text-align: center;
                        text-decoration: none;
                        text-transform: uppercase;
                        transition: 
            box-shadow .2s cubic-bezier(0.4, 0, 1, 1),
            background-color .2s cubic-bezier(0.4, 0, 0.2, 1),
            color .2s cubic-bezier(0.4, 0, 0.2, 1)
        ;
                        vertical-align: middle;
                        white-space: nowrap;
                    }
                    button:hover {
                        box-shadow: 0 3px 6px rgba(0, 0, 0, .16), 0 3px 6px rgba(0, 0, 0, .23);
                    }
                    button:active {
                        box-shadow: 0 3px 6px rgba(0, 0, 0, .16), 0 3px 6px rgba(0, 0, 0, .23);
                    }
                    button:active:hover {
                        box-shadow: 0 10px 20px rgba(0, 0, 0, .19), 0 6px 6px rgba(0, 0, 0, .23);
                        background: rgba(0, 0, 0, 0.12);
                    }
                    button:disabled {
                        pointer-events: none !important;
                        color: #a0a5aa !important;
                    }
                    button.focusable:focus {
                        outline: 1.1px auto rgba(19, 124, 189, 0.5);
                        outline-offset: 2px;
                    }
                    button.focusable:focus:active {
                        outline: 0;
                    }
                    .btn-flat,
                    .btn-flat.btn-primary {
                        background: rgba(0, 0, 0, 0);
                        box-shadow: none !important;
                    }
                    .btn-flat:hover,
                    .btn-flat.btn-primary:hover {
                        background: rgba(158, 158, 158, 0.2);
                    }
                    .btn-flat:active,
                    .btn-flat.btn-primary:active {
                        background: rgba(13, 13, 13, 0.12);
                    }
                    .btn-flat:hover:active,
                    .btn-flat.btn-primary:hover:active {
                        background: rgba(0, 0, 0, 0.12);
                    }
                    .btn-flat.btn-primary {
                        color: #0085ba;
                    }
                    .btn-primary {
                        background: #0085ba;
                        color: white;
                    }
                    .btn-primary:active:hover {
                        background: #007aab;
                    }
                    .btn-primary:disabled {
                        opacity: 0.6;
                        box-shadow: none;
                        color: white !important;
                    }
                    .btn-icon {
                        border-radius: 50%;
                        font-size: 24px;
                        height: 32px;
                        margin: auto 0;
                        min-width: 32px;
                        width: 32px;
                        padding: 0;
                        line-height: normal;
                    }
                    .btn-icon:hover {
                        color: #555555;
                    }

                    .dashicons {
                        line-height: 32px !important;
                        width: 32px;
                        height: 32px;
                    }
                    .dashicons-migrate {
                        transform: rotateY(180deg);
                        padding-left: 5px;
                    }
                    .dashicons-minus {
                        font-weight: 900;
                    }
                    .dashicons-menu,
                    .dashicons-no-alt {
                        font-size: 25px;
                    }
                    .dashicons-plus {
                        padding-top: 2px;
                    }
                
    </style>
  </div>
</Button>
`;

exports[`<Button /> Text Button with all variants 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -1,14 +1,17 @@
 <Button
+  flat={true}
+  focusable={true}
   label=\\"Click Me!\\"
+  primary={true}
   type=\\"button\\"
 >
   <div>
     <button
       aria-describedby={undefined}
       aria-label={undefined}
-      className=\\"\\"
+      className=\\"focusable btn-primary btn-flat\\"
       onClick={undefined}
       onMouseEnter={undefined}
       onMouseLeave={undefined}
       type=\\"button\\"
     >"
`;

exports[`<Button /> Text Button with defaults 1`] = `
<Button
  label="Click Me!"
  type="button"
>
  <div>
    <button
      aria-describedby={undefined}
      aria-label={undefined}
      className=""
      onClick={undefined}
      onMouseEnter={undefined}
      onMouseLeave={undefined}
      type="button"
    >
      Click Me!
    </button>
    <style
      jsx={true}
    >
      
                    button {
                        -webkit-appearance: none;
                        -webkit-touch-callout: none;
                        -webkit-user-select: none;
                        -khtml-user-select: none;
                        -moz-user-select: none;
                        -ms-user-select: none;
                        user-select: none;
                        background: white;
                        box-sizing: border-box;
                        color: #555555;
                        background-image: none;
                        border: none;
                        border-radius: 2px;
                        box-shadow: 0 1px 3px rgba(0, 0, 0, .12), 0 1px 2px rgba(0, 0, 0, .24);
                        cursor: pointer;
                        display: inline-block;
                        font-size: 13px;
                        font-weight: 500;
                        height: 36px;
                        line-height: 36px;
                        min-width: 64px;
                        outline: 0;
                        padding: 0 16px;
                        text-align: center;
                        text-decoration: none;
                        text-transform: uppercase;
                        transition: 
            box-shadow .2s cubic-bezier(0.4, 0, 1, 1),
            background-color .2s cubic-bezier(0.4, 0, 0.2, 1),
            color .2s cubic-bezier(0.4, 0, 0.2, 1)
        ;
                        vertical-align: middle;
                        white-space: nowrap;
                    }
                    button:hover {
                        box-shadow: 0 3px 6px rgba(0, 0, 0, .16), 0 3px 6px rgba(0, 0, 0, .23);
                    }
                    button:active {
                        box-shadow: 0 3px 6px rgba(0, 0, 0, .16), 0 3px 6px rgba(0, 0, 0, .23);
                    }
                    button:active:hover {
                        box-shadow: 0 10px 20px rgba(0, 0, 0, .19), 0 6px 6px rgba(0, 0, 0, .23);
                        background: rgba(0, 0, 0, 0.12);
                    }
                    button:disabled {
                        pointer-events: none !important;
                        color: #a0a5aa !important;
                    }
                    button.focusable:focus {
                        outline: 1.1px auto rgba(19, 124, 189, 0.5);
                        outline-offset: 2px;
                    }
                    button.focusable:focus:active {
                        outline: 0;
                    }
                    .btn-flat,
                    .btn-flat.btn-primary {
                        background: rgba(0, 0, 0, 0);
                        box-shadow: none !important;
                    }
                    .btn-flat:hover,
                    .btn-flat.btn-primary:hover {
                        background: rgba(158, 158, 158, 0.2);
                    }
                    .btn-flat:active,
                    .btn-flat.btn-primary:active {
                        background: rgba(13, 13, 13, 0.12);
                    }
                    .btn-flat:hover:active,
                    .btn-flat.btn-primary:hover:active {
                        background: rgba(0, 0, 0, 0.12);
                    }
                    .btn-flat.btn-primary {
                        color: #0085ba;
                    }
                    .btn-primary {
                        background: #0085ba;
                        color: white;
                    }
                    .btn-primary:active:hover {
                        background: #007aab;
                    }
                    .btn-primary:disabled {
                        opacity: 0.6;
                        box-shadow: none;
                        color: white !important;
                    }
                    .btn-icon {
                        border-radius: 50%;
                        font-size: 24px;
                        height: 32px;
                        margin: auto 0;
                        min-width: 32px;
                        width: 32px;
                        padding: 0;
                        line-height: normal;
                    }
                    .btn-icon:hover {
                        color: #555555;
                    }

                    .dashicons {
                        line-height: 32px !important;
                        width: 32px;
                        height: 32px;
                    }
                    .dashicons-migrate {
                        transform: rotateY(180deg);
                        padding-left: 5px;
                    }
                    .dashicons-minus {
                        font-weight: 900;
                    }
                    .dashicons-menu,
                    .dashicons-no-alt {
                        font-size: 25px;
                    }
                    .dashicons-plus {
                        padding-top: 2px;
                    }
                
    </style>
  </div>
</Button>
`;

exports[`<Button /> Text Button with flat 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -1,14 +1,15 @@
 <Button
+  flat={true}
   label=\\"Click Me!\\"
   type=\\"button\\"
 >
   <div>
     <button
       aria-describedby={undefined}
       aria-label={undefined}
-      className=\\"\\"
+      className=\\"btn-flat\\"
       onClick={undefined}
       onMouseEnter={undefined}
       onMouseLeave={undefined}
       type=\\"button\\"
     >"
`;

exports[`<Button /> Text Button with focusable 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -1,14 +1,15 @@
 <Button
+  focusable={true}
   label=\\"Click Me!\\"
   type=\\"button\\"
 >
   <div>
     <button
       aria-describedby={undefined}
       aria-label={undefined}
-      className=\\"\\"
+      className=\\"focusable\\"
       onClick={undefined}
       onMouseEnter={undefined}
       onMouseLeave={undefined}
       type=\\"button\\"
     >"
`;

exports[`<Button /> Text Button with primary 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -1,14 +1,15 @@
 <Button
   label=\\"Click Me!\\"
+  primary={true}
   type=\\"button\\"
 >
   <div>
     <button
       aria-describedby={undefined}
       aria-label={undefined}
-      className=\\"\\"
+      className=\\"btn-primary\\"
       onClick={undefined}
       onMouseEnter={undefined}
       onMouseLeave={undefined}
       type=\\"button\\"
     >"
`;

exports[`<Button /> Text Button with tooltip 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -1,17 +1,68 @@
 <Button
   label=\\"Click Me!\\"
+  tooltip={
+    Object {
+      \\"position\\": \\"top\\",
+      \\"text\\": \\"Hello world\\",
+    }
+  }
   type=\\"button\\"
 >
   <div>
+    <Tooltip
+      active={false}
+      id=\\"Click_Me!\\"
+      text=\\"Hello world\\"
+      transform=\\"\\"
+    >
+      <div
+        className=\\"abt-tooltip\\"
+        id=\\"Click_Me!\\"
+        role=\\"tooltip\\"
+        style={
+          Object {
+            \\"transform\\": \\"\\",
+          }
+        }
+      >
+        Hello world
+        <style
+          jsx={true}
+        >
+          
+                    .abt-tooltip {
+                        pointer-events: none;
+                        transform-origin: top center;
+                        z-index: 1000;
+                        background: rgba(66, 66, 66, 0.9);
+                        border-radius: 2px;
+                        color: white;
+                        display: inline-block;
+                        visibility: hidden;
+                        font-size: 10px;
+                        font-weight: 500;
+                        line-height: 14px;
+                        position: absolute;
+                        padding: 8px;
+                        text-align: center;
+                        max-width: 170px;
+                    }
+                    .abt-tooltip--active {
+                        visibility: visible;
+                    }
+                
+        </style>
+      </div>
+    </Tooltip>
     <button
-      aria-describedby={undefined}
+      aria-describedby=\\"Click_Me!\\"
       aria-label={undefined}
       className=\\"\\"
       onClick={undefined}
-      onMouseEnter={undefined}
-      onMouseLeave={undefined}
+      onMouseEnter={[Function anonymous]}
+      onMouseLeave={[Function anonymous]}
       type=\\"button\\"
     >
       Click Me!
     </button>
     <style"
`;
